From 72719edd270f935f9054e286e257cd8e7ed2698c Mon Sep 17 00:00:00 2001
From: Fushan Wen <qydwhotmail@gmail.com>
Date: Mon, 19 May 2025 19:56:15 +0800
Subject: [PATCH] components/sessionprivate: fix a potential crash in
 SessionsModel

The problem is that std::function is passed by reference instead of
copy, which might become dangling in the finished slot.


(cherry picked from commit 3ed6ee211bbad1f5d76afa1f30844bebad17bed4)

Co-authored-by: Fushan Wen <qydwhotmail@gmail.com>

(cherry picked from commit 995b597b24ca96aaea81f71781de8ddf60adf5e1)

Co-authored-by: Fushan Wen <qydwhotmail@gmail.com>
---
 components/sessionsprivate/sessionsmodel.cpp | 4 ++--
 components/sessionsprivate/sessionsmodel.h   | 2 +-
 2 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/components/sessionsprivate/sessionsmodel.cpp b/components/sessionsprivate/sessionsmodel.cpp
index b3f66f4d2c3..20bda2b6c12 100644
--- a/components/sessionsprivate/sessionsmodel.cpp
+++ b/components/sessionsprivate/sessionsmodel.cpp
@@ -187,11 +187,11 @@ void SessionsModel::reload()
     }
 }
 
-void SessionsModel::checkScreenLocked(const std::function<void(bool)> &cb)
+void SessionsModel::checkScreenLocked(std::function<void(bool)> &&cb)
 {
     auto reply = m_screensaverInterface->GetActive();
     QDBusPendingCallWatcher *watcher = new QDBusPendingCallWatcher(reply, this);
-    QObject::connect(watcher, &QDBusPendingCallWatcher::finished, this, [cb](QDBusPendingCallWatcher *watcher) {
+    QObject::connect(watcher, &QDBusPendingCallWatcher::finished, this, [cb = std::move(cb)](QDBusPendingCallWatcher *watcher) {
         QDBusPendingReply<bool> reply = *watcher;
         if (!reply.isError()) {
             cb(reply.value());
diff --git a/components/sessionsprivate/sessionsmodel.h b/components/sessionsprivate/sessionsmodel.h
index 81084748b16..bc580d7406d 100644
--- a/components/sessionsprivate/sessionsmodel.h
+++ b/components/sessionsprivate/sessionsmodel.h
@@ -91,7 +91,7 @@ Q_SIGNALS:
     void aboutToLockScreen();
 
 private:
-    void checkScreenLocked(const std::function<void(bool)> &cb);
+    void checkScreenLocked(std::function<void(bool)> &&cb);
 
     KDisplayManager m_displayManager;
 
-- 
GitLab

